agent/consul/state/acl.go

```diff
@@ -25,10 +62,186 @@ func (s *Restore) ACL(acl *structs.ACL) error {
 	if err := indexUpdateMaxTxn(s.tx, acl.ModifyIndex, "acls"); err != nil {
 		return fmt.Errorf("failed updating index: %s", err)
 	}
+	return nil
+}
+
+// ACLBootstrap is used to pull the ACL bootstrap info from the snapshot. This
+// might return nil, in which case nothing should be saved to the snapshot.
+func (s *Snapshot) ACLBootstrap() (*structs.ACLBootstrap, error) {
+	existing, err := s.tx.First("acls-bootstrap", "id")
+	if err != nil {
+		return nil, fmt.Errorf("failed acl bootstrap lookup: %s", err)
+	}
+	if existing != nil {
+		return existing.(*structs.ACLBootstrap), nil
+	}
+	return nil, nil
+}
+
+// ACLBootstrap is used to restore the ACL bootstrap info from the snapshot.
+func (s *Restore) ACLBootstrap(bs *structs.ACLBootstrap) error {
+	if err := s.tx.Insert("acls-bootstrap", bs); err != nil {
+		return fmt.Errorf("failed updating acl bootstrap: %v", err)
+	}
+	return nil
+}
+
+// ACLBootstrapInit is used to perform a scan for existing tokens which will
+// decide whether bootstrapping is allowed for a cluster. This is initiated by
+// the leader when it steps up, if necessary. This is because the state store
+// snapshots would become incompatible with older agents if we added this on
+// the fly, so we rely on the leader to determine a safe time to add this so
+// we can start tracking whether bootstrap is enabled. This will return an
+// error if bootstrap is already initialized.
```

Comment:
What's the reason for having this init step as its own separate thing, rather than just checking for the nonce/existing management tokens and server versions when we get a request to bootstrap?